### DISCUSSION TITLE
Feature request to invoke SIP CANCEL.

### DISCUSSION THREAD

**Discussion by imino123** [tayler6000/pyVoIP#114] [Ideas]
I'd like to kindly suggest a new feature for the pyVoIP library. It would be incredibly useful to have the ability to terminate a SIP session while it's in the 'DIALLING' state, allowing calls to be ended even if they're still ringing. To accomplish this, the 'CANCEL' SIP request could be utilised.

Curently pyVoIP 1.6.4 supports ending calls once they've reached the 'ANSWERED' state through the "hangup()" or "stop()" methods. Introducing this enhancement would significantly improve the library's versatility and user experience.

**Comment by oliverh57** [comment]
Did you have any luck with this or find a workaround - I'm currently looking into the same thing at the moment.

**Reply by imino123** [reply]
Unfortunately, I haven't been able to make any progress on a workaround for this.

**Reply by oliverh57** [reply]
I have created a fork with the cancel request implemented

https://github.com/oliverh57/pyVoIP-Cancel--Request

Example of how to use:
current_call = phone.call("111")  #creates a call
current_call.cancel_request()  #hangsup

**Reply by imino123** [reply]
Do you have a sample script i can use that makes a call then cancels request 😅
i'm a noob and can't get this to work

**Reply by imino123** [reply]
Great work on the cancel feature.

i managed to get it to work. It was a case of importing the 'VoIPCAll' library that was the issue.

I noticed that the changes made in the 'hangup()' function source code breaks the SIP 'BYE' requests of the 'hangup()' and 'stop()' functions, preventing the hangup of a call during answered state.

To fix this, I reverted back to the original code with the use of `self.sip.bye(self.request)` instead of `self.sip.cancel(self.request)` in the hangup() function's source code.

I also modified cancel_request() to handle the case when the call is already answered.

So this fork should now have the ability to hangup a call during call ringing and call answered state:

https://github.com/imino123/pyVoIP-custom

**Reply by Masl0v** [reply]
Спасибо большое! Только благодаря  этому я смог завершить свой проект автообзвона клиентов о готовности заказа. Спасибо! Спасибо! Спасибо!
